### DISCUSSION TITLE
Learning about Firefox functionalities

### DISCUSSION THREAD

**Discussion by jsm4228** [mozilla-mobile/firefox-ios#7790] [Q&A]
Hello,

I am trying to learn about how two functionalities work on Firefox iOS.

a) When a user inputs a URL and hits ENTER, the browser retrieves the page and renders it
on the browser
b) When a user downloads a document, the document is retrieved from the server, the
progress of the download is shown to the user and finally the downloaded file is saved
to the device.

Could someone direct me to where I can find the code that performs these functions? Thank you!

**Comment by smup2** [comment]
hello,

were you able to solve the above question?

**Reply by lmarceau** [reply]
Hello, you probably want to start looking with `Tab.swift`. Each tab has a webview, which loads requests (so with `func loadRequest(_ request: URLRequest)`). For download you should look into `DownloadQueue`, `DownloadContentScript` and `BrowserViewController+DownloadQueueDelegate`.
